frame/staking-rewards/src/lib.rs

```diff
@@ -840,80 +878,65 @@ pub mod pallet {
+
+			match do_reward_accumulation::<T>(reward, &pool_account, now_seconds) {
+				Ok(()) => {},
+				Err(BackToTheFuture) => {
+					Self::deposit_event(Event::<T>::RewardAccumulationHookError {
 						pool_id,
 						asset_id: reward.asset_id,
+						error: RewardAccumulationHookError::BackToTheFuture,
+					});
+				},
+				Err(RewardsPotEmpty) => {
+					Self::deposit_event(Event::<T>::RewardAccumulationHookError {
+						pool_id,
+						asset_id: reward.asset_id,
+						error: RewardAccumulationHookError::RewardsPotEmpty,
 					});
-					reward
 				},
-				Err(MaxRewardsAccumulated(reward)) => {
+				Err(MaxRewardsAccumulated) => {
 					Self::deposit_event(Event::<T>::MaxRewardsAccumulated {
 						pool_id,
 						asset_id: reward.asset_id,
 					});
-					reward
 				},
-				Err(MaxRewardsAccumulatedPreviously(reward)) => reward,
+				Err(MaxRewardsAccumulatedPreviously) => {},
 			}
 		}
 
 		pub(crate) fn acumulate_rewards_hook() -> Weight {
 			let now_seconds = T::UnixTime::now().as_secs();
 			let unix_time_now_weight = T::WeightInfo::unix_time_now();
 
-			let updated_pools = RewardPools::<T>::iter()
-				.into_iter()
-				.map(|(pool_id, reward_pool)| {
-					let updated_rewards = reward_pool
-						.rewards
-						.into_iter()
-						.map(|(asset_id, reward)| {
-							(
-								asset_id,
-								Self::reward_accumulation_hook_reward_update_calculation(
-									pool_id,
-									reward,
-									now_seconds,
-								),
-							)
-						})
-						.try_collect()
-						// SAFETY(benluelo): No elements were added to the BTreeMap; the only
-						// operation was `.map()`. This expect call will be unnecessary once this PR
-						// is merged and we update to whatever version it's included in:
-						// https://github.com/paritytech/substrate/pull/11869
-						.expect("no elements were added; qed;");
-
-					(pool_id, RewardPool { rewards: updated_rewards, ..reward_pool })
-				})
-				// NOTE(benluelo): As per these docs on `StorageMap::iter`:
-				// https://github.com/paritytech/substrate/blob/cac91f59b9e3fb8fd59842c023f87b4206931993/frame/support/src/storage/types/map.rs,
-				// "If you alter the map while doing this, you'll get undefined results."
-				// hence the double iteration.
-				.collect::<Vec<_>>();
-
 			let mut total_weight = unix_time_now_weight;
 
-			for (pool_id, reward_pool) in updated_pools {
-				// 128 bit platforms don't exist as of writing this so this usize -> u64 cast should
-				// be ok
-				let number_of_rewards_in_pool = reward_pool.rewards.len() as u64;
+			RewardPools::<T>::translate(|pool_id, mut reward_pool: RewardPoolOf<T>| {
+				for (_asset_id, reward) in &mut reward_pool.rewards {
+					Self::reward_accumulation_hook_reward_update_calculation(
+						pool_id,
+						reward,
+						now_seconds,
+					);
+				}
 
-				RewardPools::<T>::insert(pool_id, reward_pool);
+				// 128 bit platforms don't exist as of writing this so this usize -> u64 cast
+				// should be ok
```

Comment:
` reward_pool.rewards` is limited `T::MaxRewardConfigsPerPool`, [T::MaxRewardConfigsPerPool](https://github.com/ComposableFi/composable/blob/main/frame/staking-rewards/src/lib.rs#L270) is `u32`.

Reply:
good point, I've updated the comment :+1: